src/main/java/com/blackduck/integration/detect/ApplicationUpdater.java

```diff
@@ -654,30 +668,114 @@ private File download(File installDirectory, HttpUrl downloadUrl, String current
     }
     
     private File handleResponse(Response response, String currentInstalledVersion, File installDirectory, HttpUrl downloadUrl) throws IOException, IntegrationException {
+        if (response.isStatusCodeSuccess()) {
+            return handleSuccess(response, currentInstalledVersion, installDirectory, downloadUrl);
+        } else if (response.getStatusCode() == HttpStatus.SC_NOT_MODIFIED) {
+            logger.info("{} Present Detect installation is up to date - skipping download.", LOG_PREFIX);
+        } else {
+            handleFailure(response, downloadUrl);
+        }
+        return null;
+    }
+
+    private File handleSuccess(Response response, String currentInstalledVersion, File installDirectory, HttpUrl downloadUrl) throws IOException, IntegrationException {
         String newVersionString = response.getHeaderValue(DOWNLOAD_VERSION_HEADER);
-        String newFileName;
-        if (newVersionString == null && (newFileName = response.getHeaderValue(DOWNLOADED_FILE_NAME)) != null) {
+        String newFileName = response.getHeaderValue(DOWNLOADED_FILE_NAME);
+
+        if (newFileName == null) {
+            newFileName = extractFileNameFromUrl(downloadUrl);
+            
+            if (newFileName == null){
+                logger.warn("Unable to determine Detect jar filename. Detect update will not occur.");
+                return null;
+            }
+        }
+        
+        if (newVersionString == null) {
             newVersionString = getVersionFromDetectFileName(newFileName);
         }
+
+        File potentialNewJar = downloadNewJarIfNeeded(response, installDirectory, newFileName, newVersionString);
+
+        if (potentialNewJar != null) {
+            newVersionString = getVersionFromJar(potentialNewJar);
+        }
+        
+        if (newVersionString == null) {
+            logger.warn("Unable to determine version of new Detect candidate. Detect update will not occur.");
```

Comment:
To align with the wording for the _newFileName == null condition, a suggestion for a tweak:
"Unable to determine Detect jar version. Detect update will not occur."